api/include/opentelemetry/logs/logger.h

```diff
@@ -43,81 +60,105 @@ class Logger
+
+  /*** Overloaded methods for KeyValueIterables ***/
+  /**
+   * The secondary base Log(...) method that all other Log(...) overloaded methods except the one
+   * above will eventually call,  in order to create a log record.
+   *
+   * Note this takes in template types for the resource and attributes fields.
+   */
+  template <class T,
+            class U,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<U>::value> * = nullptr>
+  void Log(core::SystemTimestamp timestamp,
+           Severity severity,
+           nostd::string_view name,
+           nostd::string_view body,
+           const T &resource,
+           const U &attributes,
+           trace::TraceId trace_id,
+           trace::SpanId span_id,
+           trace::TraceFlags trace_flags) noexcept
   {
-    // Set severity to the default then call log(Severity, String message) method
-    log(Severity::kDefault, message);
+    Log(timestamp, severity, name, body, common::KeyValueIterableView<T>(resource),
+        common::KeyValueIterableView<U>(attributes), trace_id, span_id, trace_flags);
   }
 
-  inline void log(Severity severity, nostd::string_view message) noexcept
+  void Log(core::SystemTimestamp timestamp,
+           Severity severity,
+           nostd::string_view name,
+           nostd::string_view body,
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> resource,
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> attributes,
+           trace::TraceId trace_id,
+           trace::SpanId span_id,
+           trace::TraceFlags trace_flags) noexcept
   {
-    // TODO: set default timestamp later (not in API)
-    log(severity, message, core::SystemTimestamp(std::chrono::system_clock::now()));
+    return this->Log(timestamp, severity, name, body,
+                     nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+                         resource.begin(), resource.end()},
+                     nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+                         attributes.begin(), attributes.end()},
+                     trace_id, span_id, trace_flags);
   }
 
-  inline void log(Severity severity,
+  /** Wrapper methods that the user could call for convenience when logging **/
+
+  // Set default values for unspecified fields, then call the base Log() method
+  inline void Log(Severity severity,
                   nostd::string_view message,
                   core::SystemTimestamp time) noexcept
   {
-    // creates a LogRecord object with given parameters, then calls log(LogRecord)
-    LogRecord r;
-    r.severity  = severity;
-    r.body      = message;
-    r.timestamp = time;
-
-    log(r);
+    this->Log(time, severity, default_name_, message, {}, {}, default_trace_id_, default_span_id_,
```

Comment:
Could the `nullptr` be passed into `Log` instead of `{}` which may involves object construction?

Reply:
Looked into this. I believe since the `KeyValueIterable` is a `const &` reference, and not a pointer itself, it cannot point to a `nullptr` and needs to point to an object with allocated memory (https://stackoverflow.com/questions/4364536/is-null-reference-possible). In this case I believe `{}` should be the best option here